bootstrap/kubeadm/internal/ignition/clc/clc.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package clc generates Ignition using Container Linux Config Transpiler.
+package clc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/template"
+
+	ignition "github.com/coreos/ignition/config/v2_3"
+	ignitionTypes "github.com/coreos/ignition/config/v2_3/types"
+	clct "github.com/flatcar-linux/container-linux-config-transpiler/config"
+	"github.com/pkg/errors"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/internal/cloudinit"
+)
+
+const (
+	clcTemplate = `---
+systemd:
+  units:
+    - name: kubeadm.service
+      enabled: true
+      contents: |
+        [Unit]
+        Description=kubeadm
+        [Service]
+        # To not restart the unit when it exits, as it is expected.
+        Type=oneshot
+        ExecStart=/etc/kubeadm.sh
+        [Install]
+        WantedBy=multi-user.target
+
+storage:
+  files:{{ range .WriteFiles }}
+  - path: {{.Path}}
+    {{ if ne .Permissions "" -}}
+    mode: {{.Permissions}}
+    {{ end -}}
+    contents:
+      inline: |
+{{.Content | Indent 8}}
+{{- end }}
+  - path: /etc/kubeadm.sh
+    mode: 0700
+    contents:
+      inline: |
+        #!/bin/bash
+        set -e
+
+        {{- range .PreKubeadmCommands }}
+        {{ . }}
+        {{- end }}
+
+        {{ .KubeadmCommand }}
+        rm /etc/kubeadm.yml
```

Comment:
is there a specific reason to remove this file? for debugging purpose is it possible to just leave it there?

Reply:
With cloud-init, this file is placed in /tmp, so it will be removed after the reboot. With Ignition, we can't write file to /tmp, so I remove it manually instead. Now that I think about it, I think we can simply move this file to /tmp after successful execution to retain the cloud-init behavior!